### PR TITLE
New Gradle Docs Fathead

### LINKED ISSUE #551: Java: Create New Gradle Docs Fathead
# Create a new Gradle Docs Fathead Instant Answer
Help us make DuckDuckGo the best search engine for programmers!

### What do I need to know?
You'll need to know how to code in **Perl**, **Python**, **Ruby**, or **JavaScript**.
fathead languages

### What am I doing?
You will write a script that scrapes or downloads the data source below, and generates an **output.txt** file containing the parsed documentation. You can learn more about Fatheads and the `output.txt` syntax [**here**](https://docs.duckduckhack.com/resources/fathead-overview.html).

**More specifically**,
For each of the Gradle Packages, we should scrape the **Interfaces**, **Classes**, **Enums**, **Exceptions**, etc.

The JavaDoc Tree looks like a good place to start: https://docs.gradle.org/current/javadoc/overview-tree.html

**Data source**: https://docs.gradle.org/current/javadoc/

**Instant Answer Page**: *to be created by developer*

### What is the Goal?
As part of our [Programming Mission](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53), we're aiming to reach 100% Instant Answer (IA) coverage for searches related to programming languages by creating new Instant Answers, and improving existing ones.

Here are some Fathead examples:

- Ruby Docs
    - [Code](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/ruby) | [Example Query](https://duckduckgo.com/?q=array+bsearch&ia=about)
- MDN CSS
    - [Code](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/mdn_css) | [Example Query](https://duckduckgo.com/?q=css+background-position&ia=about)

  css background-position search

[See more related Instant Answers](https://duck.co/ia?repo=fathead)

## Get Started
- [ ] 1) Claim this issue by commenting below
- [ ] 2) Review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/CONTRIBUTING.md)
- [ ] 3) [Set up your development environment](https://docs.duckduckhack.com/welcome/setup-dev-environment.html), and fork this repository
- [ ] 4) Create a new Instant Answer Page: https://duck.co/ia/new_ia (then let us know, here!)
- [ ] 5) Create the Fathead
- [ ] 6) Create a Pull Request
- [ ] 7) Ping @marti1125 for a review

## Resources
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) to ask questions
- Join the [DuckDuckHack Forum](https://forum.duckduckhack.com/) to discuss project planning and Instant Answer metrics
- Read the [DuckDuckHack Documentation](https://docs.duckduckhack.com/) for technical help

### PR DESCRIPTION
**Pull request by ahetawal-p**
## Description of new Instant Answer, or changes
For each of the Gradle Packages provides an easy look up for the Interfaces, Classes, Enums, Exceptions, etc.

## People to notify
@marti1125 please review.

Fixes #551 

## Testing & Review
To be completed by Language Leader (or DDG Staff)

**Pull Request**
- [x] Title follows correct format (Specifies Instant Answer + Purpose)
- [x] Description contains a valid Instant Answer Page Link

**Instant Answer Page**
- [x] Instant Answer page is correctly filled out and contains:
    - One topic for the Search Space Language (Java, Python, Scala, Ruby, etc.)
    - One topic from: Reference, Help, Libraries, Tools
        - Documentation Fatheads are considered "Reference"
    - Description, Data source, and 2+ example queries
    - Perl Module (e.g. "DDG::Fathead::PerlDoc" -- we only need a name, not an actual file)
    - Source Name (for "More at <source_name>" link)
    - Source Domain (must contain http:// or https:// -- can be the same as Data Source)
    - Source Info (used as Subtitle for each Article -- usually matches the IA Name)
    - 'Skip Abstract' is checked off
    - Source ID (ping @moollaza to assign one, once Fathead is ready for Beta deploy)
- [ ] Fathead Tests are passing (run `$ duckpan test <fathead_id>`)
    - Tester should report any failures

**Code**
- [x] Uniformly indented, well commented
- [ ] Fetch.sh and Parse.sh run without errors
- [x] Output contains no blank lines, or multi-line entries

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/gradle_docs

### CODE REVIEW

**Comment by daxtheduck** [comment]
@ahetawal-p Thanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the [best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=TaskExecutionRequest)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/gradle_docs_mobile.png?rXtOJSdCqN)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=TaskExecutionRequest)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/gradle_docs_mobile.png?pbJH8750x9)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=TaskExecutionRequest)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/gradle_docs_mobile.png?TZfwQqos8T)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=TaskExecutionRequest)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/gradle_docs_mobile.png?OOWb3cmxSB)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=TaskExecutionRequest)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/gradle_docs_mobile.png?VbpL72PK4x)

**Comment by daxtheduck** [comment]
## [Gradle Docs](https://duck.co/ia/view/gradle_docs)

**Description**: For each of the Gradle Packages provides an easy look up for the Interfaces, Classes, Enums, Exceptions, etc.

**Example Query**: [TaskExecutionRequest](https://beta.duckduckgo.com/?q=TaskExecutionRequest), [ComponentArtifactIdentifier](https://beta.duckduckgo.com/?q=ComponentArtifactIdentifier)

**Tab Name**: About

**Source**: https://docs.gradle.org/current/javadoc/overview-summary.html

*These are the important fields from the IA page.  Please check these for errors or missing information and update the [IA page](https://duck.co/ia/view/gradle_docs)*

---
*This is an automated message which will be updated as changes are made to the [IA page](https://duck.co/ia/view/gradle_docs)*

**Comment by ahetawal-p** [comment]
@tagawa I was wondering if my latest commits can be deployed to the test site. I tried multiple commits but it doesnt seem to trigger an auto deploy?

**Comment by tagawa** [comment]
@ahetawal-p That's true. I've seen a similar problem in a separate issue. I'm not able to deploy Fatheads myself but have reported this internally. Thanks for highlighting it.

**Comment by tagawa** [comment]
Hi @ahetawal-p. This should be working on the beta server now for testing: https://beta.duckduckgo.com/?q=TaskExecutionRequest

**Comment by ahetawal-p** [comment]
@marti1125 whenever you get time, could you please take a look at this PR...

**Comment by moollaza** [comment]
@ahetawal-p can you please add a `trigger_words.txt` file?

```
gradle
gradle interface 
interface
public interface
...
```

then, those worlds will be able to come before or after any of the article titles in the output.txt to trigger this Fathead. e.g. "public interface TaskExecutionRequest" or "gradle TaskExecutionRequest"

File: lib/fathead/gradle_docs/parse.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                    // }
+
+                    // Removing <a> tags for now.
+                    $(this).replaceWith($(this).html());
+                });
+
+                //Wrap text nodes with <p>
+                //detailRowElement.contents().filter(parse_util.isTextNode).wrap(P_DOM_WRAPPER).end();
+
+                let detailNode = "<section class=\"prog__container\">"
```
Comment (moollaza) [inline lib/fathead/gradle_docs/parse.js]: nit: why not use single quotes so we don't need to escape the inner double quotes?

File: lib/fathead/gradle_docs/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+node_modules
```
Comment (ahetawal-p) [inline lib/fathead/gradle_docs/.gitignore]: Added this for when working with just this folder. Seems the top level doesn't work when i try to push from this folder.

**Comment by ahetawal-p** [comment]
@moollaza take a look now, updated after code review comments.

**Comment by ahetawal-p** [comment]
@duckduckgo/duckduckhack-contributors can someone please take a look at this PR. If its good to be merged ?

File: lib/fathead/gradle_docs/parse.js [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                        callback(null, url);
+                        },function (err) {
+                            console.error("%s; %s", err.message, url);
+                            console.log("%j", err.res.statusCode);
+                            callback(err, url);
+                        });
+                },function(err, results) {
+                    if(err){
+                        console.error(err);
+                        console.err("Gradle Fathead Processing Errored out..." + new Date());
```
Comment (atilacamurca) [inline lib/fathead/gradle_docs/parse.js]: isn't it suppose to be `console.error`?

File: lib/fathead/gradle_docs/parse_util.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+'use strict';
```
Comment (atilacamurca) [inline lib/fathead/gradle_docs/parse_util.js]: can you proper format this file? That way will be more easy to read and review :)

**Comment by moollaza** [comment]
@ahetawal-p Thanks! This LGTM :+1:

**Comment by daxtheduck** [comment]
@ahetawal-p this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!

**Comment by moollaza** [comment]
Thanks for reviewing @marti1125 👍

**Pull request merged**